rfcs/007-background-workers.md

```diff
@@ -0,0 +1,252 @@
+
+Active queue:
+ - `("couch_workers", "active", WorkerType, Worker, JobId) = (JobState, JobInfo, JobOpts)`
+
+Worker registration and health:
+ - `("couch_workers", WorkerType, "workers_vs") = Versionstamp`
+ - `("couch_workers", WorkerType, "workers", WorkerId) = WorkerOpts`
+ - `("couch_workers", WorkerType, "health", WorkerId) = (Versionstamp, Timestamp, WorkerTimeout)`
+
+### Data model fields
+
+- `JobType`: The job type. It MAY be ``replication`` or ``indexing``, however,
+  specific types are not in scope of this document.
+
+- `JobId` : MUST unique identify a job in a cluster. It MAY be a `UUID` or
+  derived from some other job parameters.
+
+- `JobState` : MUST be one of `pending`, `running`, or `completed`
+
+- `Priority` : MAY be any key which allows sorting jobs, such as a timestamp,
+  or a tag.
+
+- `WorkerId` : MUST uniquely identify a worker in a cluster. It MAY be
+  generated each time the worker is restarted, or it could be persisted across
+  worker restarts, as long as it is unique across the whole cluster.
+
+- `CancelReq` : A boolean flag indicating a request to the worker to cancel the
+  job.
+
+- `JobOps` : Type-specific job options. Represented as a JSON object. This MAY
+  contain fields like `"source"`, `"target"`, `"dbname"` etc.
+
+- `JobInfo` : A per-job info object. Represented as JSON. This object will
+  contain details about a job's current state. It MAY have fields such as
+  "update_seq", "error_count", etc.
+
+
+## Job Lifecyle
+
+New jobs are posted to the main jobs table along with an entry in the pending
+queue. Priority key assignment is type specific. `WorkerId` is set to `nil`
+initially.
+
+Workers will monitor the pending queue with their matching `JobType`. That is,
+`"replication"` workers will monitor the `"replication"` pending queue,
+indexing workers the `"indexing"` queue, etc. If there are jobs ready to run,
+they will attempt to grab one or more jobs from the pending queue and assign it
+to themselves for execution. They do that by setting the `WorkerId` in the jobs
+table entry to their worker ID, removing the job from pending queue and create
+a new entry in their own `"active"` jobs area.
+
+When a job finishes running, either because it was a successful completion or a
+terminal failure, the worker MUST remove the job from its active queue and
+clear its `WorkerId` field. Then, based on type-specific behavior, it MUST do
+either one of:
+
+ - Update the `JobState` to `"completed"` and leave the job in the jobs table.
+   It would be up to the job creation logic to inspect and remove it.
+ - Delete the job from the system altogether. This may be useful when running
+   _replicate jobs, for example.
+
+
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+to `true`. Then they MUST wait for the job to stop running. If the job is not
+running, they may directly remove the job from the pending and the main jobs
+table.
```

Comment:
What happens in the case of very long running jobs? e.g continuous replication or indexing a very large db. If we get into a situation where the replication queue is has more continuous replications and than workers, some jobs will never be processed. I think we would have to look at having the workers periodically check the queue and decide if they need to pause their job, add it back to the pending queue with new options and pick another job.

Reply:
Exactly. Yeah, in case of replications something like that might be needed and we might preserve some of the previous scheduler features, like you pointed out:

```
  If there is a pending backlog:
     Stop `$max_churn` number of continuous jobs put them back on the queue
     Pick up `$max_churn` number of jobs from the queue and run those 
```

Good idea, I'll add this to the RFC.